src/redisraft.c

```diff
@@ -550,6 +667,75 @@ static void handleRedisCommand(RedisRaftCtx *rr,
     raft_entry_release(entry);
 }
 
+static void handleRedisCommand(RedisRaftCtx *rr,
+                               RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
+{
+    /* update cmds array with the client's saved "asking" state */
+    cmds->asking = getAskingState(rr, ctx);
+    setAskingState(rr, ctx, false);
+
+    /* Handle intercepted commands. We do this also on non-leader nodes or if we
+     * don't have a leader, so it's up to the commands to check these conditions
+     * if they have to.
+     */
+    if (handleInterceptedCommands(rr, ctx, cmds)) {
+        return;
+    }
+
+    /* Check if MULTI/EXEC bundling is required. */
+    if (MultiHandleCommand(rr, ctx, cmds)) {
+        return;
+    }
+
+    handleRedisCommandAppend(rr, ctx, cmds);
+}
+
+static void redirectCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader)
+{
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+
+    /* One anomaly here is that may redirect a client to the leader even for
+     * commands have no keys, which is something Redis Cluster never does. We
+     * still need to consider how this impacts clients which may not expect it.
+     */
+    replyRedirect(ctx, 0, &leader->addr);
+}
+
+static void handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds)
+{
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+    RedisModule_Assert(cmds);
+
+    /* reply ASK for ASKING state commands */
+    if (cmds->asking) {
+        int slot;
+        if (HashSlotCompute(rr, cmds, &slot) != RR_OK) {
+            replyCrossSlot(ctx);
+            return;
+        }
+
+        if (slot == -1) {
+            /* ASKING mode requests should always have keys */
+            RedisModule_ReplyWithError(ctx, "ERR cmd without keys in asking mode");
+            return;
+        }
+
+        replyAsk(rr, ctx, (unsigned int) slot);
```

Comment:
If I remember correctly, we want to redirect to the leader here. 

1- We can change replyAsk() function like this:

```c

void replyAsk(RedisModuleCtx *ctx, unsigned int slot, NodeAddr *addr)
{
    char buf[sizeof(addr->host) + 256];

    snprintf(buf, sizeof(buf), "ASK %u %s:%u", slot, addr->host, addr->port);
    RedisModule_ReplyWithError(ctx, buf);
}
```

2- We can move this part down below to `validateRaftRedisCommandArray()` where we make another `replyAsk()` call:

```c
    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
    if (!sg) {
        RedisModule_ReplyWithError(ctx, "ERR no importing shard group to ask");
        return;
    }
```
3- Maybe we can add a test, send "asking" to a follower, expect an "ask" response with leader's address.

Reply:
https://github.com/RedisLabs/redisraft/pull/427